sympy/polys/matrices/domainmatrix.py

```diff
@@ -444,8 +596,13 @@ def add(A, B):
             raise ShapeError("shape")
         if A.domain != B.domain:
             raise ValueError("domain")
+        if isinstance(A.rep, list) != isinstance(B.rep, list):
+            A = A.to_dense()
+            B = B.to_dense()
```

Comment:
The point of adding the dense/sparse unification logic in `unify` is so that we don't need to do it in each other method.

Reply:
I'll remove it from other methods